internal/batches/executor/executor.go

```diff
@@ -69,83 +68,72 @@ type NewExecutorOpts struct {
 	IsRemote         bool
 	GlobalEnv        []string
 	ForceRoot        bool
+	FailFast         bool
 
 	BinaryDiffs bool
+	Context     context.Context
 }
 
 type executor struct {
 	opts NewExecutorOpts
 
-	par           *parallel.Run
+	workPool      *pool.ResultContextPool[*taskResult]
 	doneEnqueuing chan struct{}
-
-	results   []taskResult
-	resultsMu sync.Mutex
 }
 
 func NewExecutor(opts NewExecutorOpts) *executor {
+	p := pool.NewWithResults[*taskResult]().WithMaxGoroutines(opts.Parallelism).WithContext(opts.Context)
+	if opts.FailFast {
+		p = p.WithCancelOnError()
+	}
 	return &executor{
 		opts: opts,
 
 		doneEnqueuing: make(chan struct{}),
-		par:           parallel.NewRun(opts.Parallelism),
+		workPool:      p,
 	}
 }
 
 // Start starts the execution of the given Tasks in goroutines, calling the
 // given taskStatusHandler to update the progress of the tasks.
-func (x *executor) Start(ctx context.Context, tasks []*Task, ui TaskExecutionUI) {
+func (x *executor) Start(tasks []*Task, ui TaskExecutionUI) {
```

Comment:
I'd suggest creating the work pool in `Start`, then saving it on the `executor` struct so you can `Wait()` it later. It's a bit of an antipattern to save a context on a struct without really good reason

Reply:
Sounds like a good change! Will tackle that before merging.